<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang xml:lang>
<head>
  <meta charset="utf-8" />
  <meta name="generator" content="pandoc" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes" />
  <meta name="author" content="Sonja Lecic" />
  <title>Parus_major_RRBS_pipeline</title>
  <style type="text/css">
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>

  <script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
  <style type="text/css">:root{--fore-color:#111;--secondary-fore-color:#444;--back-color:#f8f8f8;--secondary-back-color:#f0f0f0;--blockquote-color:#f57c00;--pre-color:#1565c0;--border-color:#aaa;--secondary-border-color:#ddd;--heading-ratio:1.19;--universal-margin:.5rem;--universal-padding:.5rem;--universal-border-radius:.125rem;--a-link-color:#0277bd;--a-visited-color:#01579b}html{font-size:16px}a,b,del,em,i,ins,q,span,strong,u{font-size:1em}html,*{font-family:-apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, Ubuntu, "Helvetica Neue", Helvetica, sans-serif;line-height:1.5;-webkit-text-size-adjust:100%}*{font-size:1rem}body{margin:0;color:var(--fore-color);background:var(--back-color)}details{display:block}summary{display:list-item}abbr[title]{border-bottom:none;text-decoration:underline dotted}input{overflow:visible}img{max-width:100%;height:auto}h1,h2,h3,h4,h5,h6{line-height:1.2;margin:calc(1.5 * var(--universal-margin)) var(--universal-margin);font-weight:500}h1 small,h2 small,h3 small,h4 small,h5 small,h6 small{color:var(--secondary-fore-color);display:block;margin-top:-.25rem}h1{font-size:calc(1rem * var(--heading-ratio) * var(--heading-ratio) * var(--heading-ratio) * var(--heading-ratio))}h2{font-size:calc(1rem * var(--heading-ratio) * var(--heading-ratio) * var(--heading-ratio))}h3{font-size:calc(1rem * var(--heading-ratio) * var(--heading-ratio))}h4{font-size:calc(1rem * var(--heading-ratio))}h5{font-size:1rem}h6{font-size:calc(1rem / var(--heading-ratio))}p{margin:var(--universal-margin)}ol,ul{margin:var(--universal-margin);padding-left:calc(2 * var(--universal-margin))}b,strong{font-weight:700}hr{box-sizing:content-box;border:0;line-height:1.25em;margin:var(--universal-margin);height:.0625rem;background:linear-gradient(to right, transparent, var(--border-color) 20%, var(--border-color) 80%, transparent)}blockquote{display:block;position:relative;font-style:italic;color:var(--secondary-fore-color);margin:var(--universal-margin);padding:calc(3 * var(--universal-padding));border:.0625rem solid var(--secondary-border-color);border-left:.375rem solid var(--blockquote-color);border-radius:0 var(--universal-border-radius) var(--universal-border-radius) 0}blockquote:before{position:absolute;top:calc(0rem - var(--universal-padding));left:0;font-family:sans-serif;font-size:3rem;font-weight:700;content:"\201c";color:var(--blockquote-color)}blockquote[cite]:after{font-style:normal;font-size:.75em;font-weight:700;content:"\aâ€” " attr(cite);white-space:pre}code,kbd,pre,samp{font-family:Menlo, Consolas, monospace;font-size:.85em}code{background:var(--secondary-back-color);border-radius:var(--universal-border-radius);padding:calc(var(--universal-padding) / 4) calc(var(--universal-padding) / 2)}kbd{background:var(--fore-color);color:var(--back-color);border-radius:var(--universal-border-radius);padding:calc(var(--universal-padding) / 4) calc(var(--universal-padding) / 2)}pre{overflow:auto;background:var(--secondary-back-color);padding:calc(1.5 * var(--universal-padding));margin:var(--universal-margin);border:.0625rem solid var(--secondary-border-color);border-left:.25rem solid var(--pre-color);border-radius:0 var(--universal-border-radius) var(--universal-border-radius) 0}sup,sub,code,kbd{line-height:0;position:relative;vertical-align:baseline}small,sup,sub,figcaption{font-size:.75em}sup{top:-.5em}sub{bottom:-.25em}figure{margin:var(--universal-margin)}figcaption{color:var(--secondary-fore-color)}a{text-decoration:none}a:link{color:var(--a-link-color)}a:visited{color:var(--a-visited-color)}a:hover,a:focus{text-decoration:underline}.container{margin:0 auto;padding:0 calc(1.5 * var(--universal-padding))}.row{box-sizing:border-box;display:flex;flex:0 1 auto;flex-flow:row wrap}.col-sm,[class^='col-sm-'],[class^='col-sm-offset-'],.row[class*='cols-sm-']>*{box-sizing:border-box;flex:0 0 auto;padding:0 calc(var(--universal-padding) / 2)}.col-sm,.row.cols-sm>*{max-width:100%;flex-grow:1;flex-basis:0}.col-sm-1,.row.cols-sm-1>*{max-width:8.33333%;flex-basis:8.33333%}.col-sm-offset-0{margin-left:0}.col-sm-2,.row.cols-sm-2>*{max-width:16.66667%;flex-basis:16.66667%}.col-sm-offset-1{margin-left:8.33333%}.col-sm-3,.row.cols-sm-3>*{max-width:25%;flex-basis:25%}.col-sm-offset-2{margin-left:16.66667%}.col-sm-4,.row.cols-sm-4>*{max-width:33.33333%;flex-basis:33.33333%}.col-sm-offset-3{margin-left:25%}.col-sm-5,.row.cols-sm-5>*{max-width:41.66667%;flex-basis:41.66667%}.col-sm-offset-4{margin-left:33.33333%}.col-sm-6,.row.cols-sm-6>*{max-width:50%;flex-basis:50%}.col-sm-offset-5{margin-left:41.66667%}.col-sm-7,.row.cols-sm-7>*{max-width:58.33333%;flex-basis:58.33333%}.col-sm-offset-6{margin-left:50%}.col-sm-8,.row.cols-sm-8>*{max-width:66.66667%;flex-basis:66.66667%}.col-sm-offset-7{margin-left:58.33333%}.col-sm-9,.row.cols-sm-9>*{max-width:75%;flex-basis:75%}.col-sm-offset-8{margin-left:66.66667%}.col-sm-10,.row.cols-sm-10>*{max-width:83.33333%;flex-basis:83.33333%}.col-sm-offset-9{margin-left:75%}.col-sm-11,.row.cols-sm-11>*{max-width:91.66667%;flex-basis:91.66667%}.col-sm-offset-10{margin-left:83.33333%}.col-sm-12,.row.cols-sm-12>*{max-width:100%;flex-basis:100%}.col-sm-offset-11{margin-left:91.66667%}.col-sm-normal{order:initial}.col-sm-first{order:-999}.col-sm-last{order:999}@media screen and (min-width: 768px){.col-md,[class^='col-md-'],[class^='col-md-offset-'],.row[class*='cols-md-']>*{box-sizing:border-box;flex:0 0 auto;padding:0 calc(var(--universal-padding) / 2)}.col-md,.row.cols-md>*{max-width:100%;flex-grow:1;flex-basis:0}.col-md-1,.row.cols-md-1>*{max-width:8.33333%;flex-basis:8.33333%}.col-md-offset-0{margin-left:0}.col-md-2,.row.cols-md-2>*{max-width:16.66667%;flex-basis:16.66667%}.col-md-offset-1{margin-left:8.33333%}.col-md-3,.row.cols-md-3>*{max-width:25%;flex-basis:25%}.col-md-offset-2{margin-left:16.66667%}.col-md-4,.row.cols-md-4>*{max-width:33.33333%;flex-basis:33.33333%}.col-md-offset-3{margin-left:25%}.col-md-5,.row.cols-md-5>*{max-width:41.66667%;flex-basis:41.66667%}.col-md-offset-4{margin-left:33.33333%}.col-md-6,.row.cols-md-6>*{max-width:50%;flex-basis:50%}.col-md-offset-5{margin-left:41.66667%}.col-md-7,.row.cols-md-7>*{max-width:58.33333%;flex-basis:58.33333%}.col-md-offset-6{margin-left:50%}.col-md-8,.row.cols-md-8>*{max-width:66.66667%;flex-basis:66.66667%}.col-md-offset-7{margin-left:58.33333%}.col-md-9,.row.cols-md-9>*{max-width:75%;flex-basis:75%}.col-md-offset-8{margin-left:66.66667%}.col-md-10,.row.cols-md-10>*{max-width:83.33333%;flex-basis:83.33333%}.col-md-offset-9{margin-left:75%}.col-md-11,.row.cols-md-11>*{max-width:91.66667%;flex-basis:91.66667%}.col-md-offset-10{margin-left:83.33333%}.col-md-12,.row.cols-md-12>*{max-width:100%;flex-basis:100%}.col-md-offset-11{margin-left:91.66667%}.col-md-normal{order:initial}.col-md-first{order:-999}.col-md-last{order:999}}@media screen and (min-width: 1280px){.col-lg,[class^='col-lg-'],[class^='col-lg-offset-'],.row[class*='cols-lg-']>*{box-sizing:border-box;flex:0 0 auto;padding:0 calc(var(--universal-padding) / 2)}.col-lg,.row.cols-lg>*{max-width:100%;flex-grow:1;flex-basis:0}.col-lg-1,.row.cols-lg-1>*{max-width:8.33333%;flex-basis:8.33333%}.col-lg-offset-0{margin-left:0}.col-lg-2,.row.cols-lg-2>*{max-width:16.66667%;flex-basis:16.66667%}.col-lg-offset-1{margin-left:8.33333%}.col-lg-3,.row.cols-lg-3>*{max-width:25%;flex-basis:25%}.col-lg-offset-2{margin-left:16.66667%}.col-lg-4,.row.cols-lg-4>*{max-width:33.33333%;flex-basis:33.33333%}.col-lg-offset-3{margin-left:25%}.col-lg-5,.row.cols-lg-5>*{max-width:41.66667%;flex-basis:41.66667%}.col-lg-offset-4{margin-left:33.33333%}.col-lg-6,.row.cols-lg-6>*{max-width:50%;flex-basis:50%}.col-lg-offset-5{margin-left:41.66667%}.col-lg-7,.row.cols-lg-7>*{max-width:58.33333%;flex-basis:58.33333%}.col-lg-offset-6{margin-left:50%}.col-lg-8,.row.cols-lg-8>*{max-width:66.66667%;flex-basis:66.66667%}.col-lg-offset-7{margin-left:58.33333%}.col-lg-9,.row.cols-lg-9>*{max-width:75%;flex-basis:75%}.col-lg-offset-8{margin-left:66.66667%}.col-lg-10,.row.cols-lg-10>*{max-width:83.33333%;flex-basis:83.33333%}.col-lg-offset-9{margin-left:75%}.col-lg-11,.row.cols-lg-11>*{max-width:91.66667%;flex-basis:91.66667%}.col-lg-offset-10{margin-left:83.33333%}.col-lg-12,.row.cols-lg-12>*{max-width:100%;flex-basis:100%}.col-lg-offset-11{margin-left:91.66667%}.col-lg-normal{order:initial}.col-lg-first{order:-999}.col-lg-last{order:999}}:root{--card-back-color:#f8f8f8;--card-fore-color:#111;--card-border-color:#ddd}.card{display:flex;flex-direction:column;justify-content:space-between;align-self:center;position:relative;width:100%;background:var(--card-back-color);color:var(--card-fore-color);border:.0625rem solid var(--card-border-color);border-radius:var(--universal-border-radius);margin:var(--universal-margin);overflow:hidden}@media screen and (min-width: 320px){.card{max-width:320px}}.card>.section{background:var(--card-back-color);color:var(--card-fore-color);box-sizing:border-box;margin:0;border:0;border-radius:0;border-bottom:.0625rem solid var(--card-border-color);padding:var(--universal-padding);width:100%}.card>.section.media{height:200px;padding:0;-o-object-fit:cover;object-fit:cover}.card>.section:last-child{border-bottom:0}@media screen and (min-width: 240px){.card.small{max-width:240px}}@media screen and (min-width: 480px){.card.large{max-width:480px}}.card.fluid{max-width:100%;width:auto}.card.warning{--card-back-color:#ffca28;--card-border-color:#e8b825}.card.error{--card-back-color:#b71c1c;--card-fore-color:#f8f8f8;--card-border-color:#a71a1a}.card>.section.dark{--card-back-color:#e0e0e0}.card>.section.double-padded{padding:calc(1.5 * var(--universal-padding))}:root{--form-back-color:#f0f0f0;--form-fore-color:#111;--form-border-color:#ddd;--input-back-color:#f8f8f8;--input-fore-color:#111;--input-border-color:#ddd;--input-focus-color:#0288d1;--input-invalid-color:#d32f2f;--button-back-color:#e2e2e2;--button-hover-back-color:#dcdcdc;--button-fore-color:#212121;--button-border-color:rgba(0,0,0,0);--button-hover-border-color:rgba(0,0,0,0);--button-group-border-color:rgba(124,124,124,0.54)}form{background:var(--form-back-color);color:var(--form-fore-color);border:.0625rem solid var(--form-border-color);border-radius:var(--universal-border-radius);margin:var(--universal-margin);padding:calc(2 * var(--universal-padding)) var(--universal-padding)}fieldset{border:.0625rem solid var(--form-border-color);border-radius:var(--universal-border-radius);margin:calc(var(--universal-margin) / 4);padding:var(--universal-padding)}legend{box-sizing:border-box;display:table;max-width:100%;white-space:normal;font-weight:700;padding:calc(var(--universal-padding) / 2)}label{padding:calc(var(--universal-padding) / 2) var(--universal-padding)}.input-group{display:inline-block}.input-group.fluid{display:flex;align-items:center;justify-content:center}.input-group.fluid>input{max-width:100%;flex-grow:1;flex-basis:0px}@media screen and (max-width: 767px){.input-group.fluid{align-items:stretch;flex-direction:column}}.input-group.vertical{display:flex;align-items:stretch;flex-direction:column}.input-group.vertical>input{max-width:100%;flex-grow:1;flex-basis:0px}[type="number"]::-webkit-inner-spin-button,[type="number"]::-webkit-outer-spin-button{height:auto}[type="search"]{-webkit-appearance:textfield;outline-offset:-2px}[type="search"]::-webkit-search-cancel-button,[type="search"]::-webkit-search-decoration{-webkit-appearance:none}input:not([type]),[type="text"],[type="email"],[type="number"],[type="search"],[type="password"],[type="url"],[type="tel"],[type="checkbox"],[type="radio"],textarea,select{box-sizing:border-box;background:var(--input-back-color);color:var(--input-fore-color);border:.0625rem solid var(--input-border-color);border-radius:var(--universal-border-radius);margin:calc(var(--universal-margin) / 2);padding:var(--universal-padding) calc(1.5 * var(--universal-padding))}input:not([type="button"]):not([type="submit"]):not([type="reset"]):hover,input:not([type="button"]):not([type="submit"]):not([type="reset"]):focus,textarea:hover,textarea:focus,select:hover,select:focus{border-color:var(--input-focus-color);box-shadow:none}input:not([type="button"]):not([type="submit"]):not([type="reset"]):invalid,input:not([type="button"]):not([type="submit"]):not([type="reset"]):focus:invalid,textarea:invalid,textarea:focus:invalid,select:invalid,select:focus:invalid{border-color:var(--input-invalid-color);box-shadow:none}input:not([type="button"]):not([type="submit"]):not([type="reset"])[readonly],textarea[readonly],select[readonly]{background:var(--secondary-back-color)}select{max-width:100%}option{overflow:hidden;text-overflow:ellipsis}[type="checkbox"],[type="radio"]{-webkit-appearance:none;-moz-appearance:none;appearance:none;position:relative;height:calc(1rem + var(--universal-padding) / 2);width:calc(1rem + var(--universal-padding) / 2);vertical-align:text-bottom;padding:0;flex-basis:calc(1rem + var(--universal-padding) / 2) !important;flex-grow:0 !important}[type="checkbox"]:checked:before,[type="radio"]:checked:before{position:absolute}[type="checkbox"]:checked:before{content:'\2713';font-family:sans-serif;font-size:calc(1rem + var(--universal-padding) / 2);top:calc(0rem - var(--universal-padding));left:calc(var(--universal-padding) / 4)}[type="radio"]{border-radius:100%}[type="radio"]:checked:before{border-radius:100%;content:'';top:calc(.0625rem + var(--universal-padding) / 2);left:calc(.0625rem + var(--universal-padding) / 2);background:var(--input-fore-color);width:0.5rem;height:0.5rem}:placeholder-shown{color:var(--input-fore-color)}::-ms-placeholder{color:var(--input-fore-color);opacity:0.54}button::-moz-focus-inner,[type="button"]::-moz-focus-inner,[type="reset"]::-moz-focus-inner,[type="submit"]::-moz-focus-inner{border-style:none;padding:0}button,html [type="button"],[type="reset"],[type="submit"]{-webkit-appearance:button}button{overflow:visible;text-transform:none}button,[type="button"],[type="submit"],[type="reset"],a.button,label.button,.button,a[role="button"],label[role="button"],[role="button"]{display:inline-block;background:var(--button-back-color);color:var(--button-fore-color);border:.0625rem solid var(--button-border-color);border-radius:var(--universal-border-radius);padding:var(--universal-padding) calc(1.5 * var(--universal-padding));margin:var(--universal-margin);text-decoration:none;cursor:pointer;transition:background 0.3s}button:hover,button:focus,[type="button"]:hover,[type="button"]:focus,[type="submit"]:hover,[type="submit"]:focus,[type="reset"]:hover,[type="reset"]:focus,a.button:hover,a.button:focus,label.button:hover,label.button:focus,.button:hover,.button:focus,a[role="button"]:hover,a[role="button"]:focus,label[role="button"]:hover,label[role="button"]:focus,[role="button"]:hover,[role="button"]:focus{background:var(--button-hover-back-color);border-color:var(--button-hover-border-color)}input:disabled,input[disabled],textarea:disabled,textarea[disabled],select:disabled,select[disabled],button:disabled,button[disabled],.button:disabled,.button[disabled],[role="button"]:disabled,[role="button"][disabled]{cursor:not-allowed;opacity:.75}.button-group{display:flex;border:.0625rem solid var(--button-group-border-color);border-radius:var(--universal-border-radius);margin:var(--universal-margin)}.button-group>button,.button-group [type="button"],.button-group>[type="submit"],.button-group>[type="reset"],.button-group>.button,.button-group>[role="button"]{margin:0;max-width:100%;flex:1 1 auto;text-align:center;border:0;border-radius:0;box-shadow:none}.button-group>:not(:first-child){border-left:.0625rem solid var(--button-group-border-color)}@media screen and (max-width: 767px){.button-group{flex-direction:column}.button-group>:not(:first-child){border:0;border-top:.0625rem solid var(--button-group-border-color)}}button.primary,[type="button"].primary,[type="submit"].primary,[type="reset"].primary,.button.primary,[role="button"].primary{--button-back-color:#1976d2;--button-fore-color:#f8f8f8}button.primary:hover,button.primary:focus,[type="button"].primary:hover,[type="button"].primary:focus,[type="submit"].primary:hover,[type="submit"].primary:focus,[type="reset"].primary:hover,[type="reset"].primary:focus,.button.primary:hover,.button.primary:focus,[role="button"].primary:hover,[role="button"].primary:focus{--button-hover-back-color:#1565c0}button.secondary,[type="button"].secondary,[type="submit"].secondary,[type="reset"].secondary,.button.secondary,[role="button"].secondary{--button-back-color:#d32f2f;--button-fore-color:#f8f8f8}button.secondary:hover,button.secondary:focus,[type="button"].secondary:hover,[type="button"].secondary:focus,[type="submit"].secondary:hover,[type="submit"].secondary:focus,[type="reset"].secondary:hover,[type="reset"].secondary:focus,.button.secondary:hover,.button.secondary:focus,[role="button"].secondary:hover,[role="button"].secondary:focus{--button-hover-back-color:#c62828}button.tertiary,[type="button"].tertiary,[type="submit"].tertiary,[type="reset"].tertiary,.button.tertiary,[role="button"].tertiary{--button-back-color:#308732;--button-fore-color:#f8f8f8}button.tertiary:hover,button.tertiary:focus,[type="button"].tertiary:hover,[type="button"].tertiary:focus,[type="submit"].tertiary:hover,[type="submit"].tertiary:focus,[type="reset"].tertiary:hover,[type="reset"].tertiary:focus,.button.tertiary:hover,.button.tertiary:focus,[role="button"].tertiary:hover,[role="button"].tertiary:focus{--button-hover-back-color:#277529}button.inverse,[type="button"].inverse,[type="submit"].inverse,[type="reset"].inverse,.button.inverse,[role="button"].inverse{--button-back-color:#212121;--button-fore-color:#f8f8f8}button.inverse:hover,button.inverse:focus,[type="button"].inverse:hover,[type="button"].inverse:focus,[type="submit"].inverse:hover,[type="submit"].inverse:focus,[type="reset"].inverse:hover,[type="reset"].inverse:focus,.button.inverse:hover,.button.inverse:focus,[role="button"].inverse:hover,[role="button"].inverse:focus{--button-hover-back-color:#111}button.small,[type="button"].small,[type="submit"].small,[type="reset"].small,.button.small,[role="button"].small{padding:calc(0.5 * var(--universal-padding)) calc(0.75 * var(--universal-padding));margin:var(--universal-margin)}button.large,[type="button"].large,[type="submit"].large,[type="reset"].large,.button.large,[role="button"].large{padding:calc(1.5 * var(--universal-padding)) calc(2 * var(--universal-padding));margin:var(--universal-margin)}:root{--header-back-color:#f8f8f8;--header-hover-back-color:#f0f0f0;--header-fore-color:#444;--header-border-color:#ddd;--nav-back-color:#f8f8f8;--nav-hover-back-color:#f0f0f0;--nav-fore-color:#444;--nav-border-color:#ddd;--nav-link-color:#0277bd;--footer-fore-color:#444;--footer-back-color:#f8f8f8;--footer-border-color:#ddd;--footer-link-color:#0277bd;--drawer-back-color:#f8f8f8;--drawer-hover-back-color:#f0f0f0;--drawer-border-color:#ddd;--drawer-close-color:#444}header{height:3.1875rem;background:var(--header-back-color);color:var(--header-fore-color);border-bottom:.0625rem solid var(--header-border-color);padding:calc(var(--universal-padding) / 4) 0;white-space:nowrap;overflow-x:auto;overflow-y:hidden}header.row{box-sizing:content-box}header .logo{color:var(--header-fore-color);font-size:1.75rem;padding:var(--universal-padding) calc(2 * var(--universal-padding));text-decoration:none}header button,header [type="button"],header .button,header [role="button"]{box-sizing:border-box;position:relative;top:calc(0rem - var(--universal-padding) / 4);height:calc(3.1875rem + var(--universal-padding) / 2);background:var(--header-back-color);line-height:calc(3.1875rem - var(--universal-padding) * 1.5);text-align:center;color:var(--header-fore-color);border:0;border-radius:0;margin:0;text-transform:uppercase}header button:hover,header button:focus,header [type="button"]:hover,header [type="button"]:focus,header .button:hover,header .button:focus,header [role="button"]:hover,header [role="button"]:focus{background:var(--header-hover-back-color)}nav{background:var(--nav-back-color);color:var(--nav-fore-color);border:.0625rem solid var(--nav-border-color);border-radius:var(--universal-border-radius);margin:var(--universal-margin)}nav *{padding:var(--universal-padding) calc(1.5 * var(--universal-padding))}nav a,nav a:visited{display:block;color:var(--nav-link-color);border-radius:var(--universal-border-radius);transition:background 0.3s}nav a:hover,nav a:focus,nav a:visited:hover,nav a:visited:focus{text-decoration:none;background:var(--nav-hover-back-color)}nav .sublink-1{position:relative;margin-left:calc(2 * var(--universal-padding))}nav .sublink-1:before{position:absolute;left:calc(var(--universal-padding) - 1 * var(--universal-padding));top:-.0625rem;content:'';height:100%;border:.0625rem solid var(--nav-border-color);border-left:0}nav .sublink-2{position:relative;margin-left:calc(4 * var(--universal-padding))}nav .sublink-2:before{position:absolute;left:calc(var(--universal-padding) - 3 * var(--universal-padding));top:-.0625rem;content:'';height:100%;border:.0625rem solid var(--nav-border-color);border-left:0}footer{background:var(--footer-back-color);color:var(--footer-fore-color);border-top:.0625rem solid var(--footer-border-color);padding:calc(2 * var(--universal-padding)) var(--universal-padding);font-size:.875rem}footer a,footer a:visited{color:var(--footer-link-color)}header.sticky{position:-webkit-sticky;position:sticky;z-index:1101;top:0}footer.sticky{position:-webkit-sticky;position:sticky;z-index:1101;bottom:0}.drawer-toggle:before{display:inline-block;position:relative;vertical-align:bottom;content:'\00a0\2261\00a0';font-family:sans-serif;font-size:1.5em}@media screen and (min-width: 768px){.drawer-toggle:not(.persistent){display:none}}[type="checkbox"].drawer{height:1px;width:1px;margin:-1px;overflow:hidden;position:absolute;clip:rect(0 0 0 0);-webkit-clip-path:inset(100%);clip-path:inset(100%)}[type="checkbox"].drawer+*{display:block;box-sizing:border-box;position:fixed;top:0;width:320px;height:100vh;overflow-y:auto;background:var(--drawer-back-color);border:.0625rem solid var(--drawer-border-color);border-radius:0;margin:0;z-index:1110;right:-320px;transition:right 0.3s}[type="checkbox"].drawer+* .drawer-close{position:absolute;top:var(--universal-margin);right:var(--universal-margin);z-index:1111;width:2rem;height:2rem;border-radius:var(--universal-border-radius);padding:var(--universal-padding);margin:0;cursor:pointer;transition:background 0.3s}[type="checkbox"].drawer+* .drawer-close:before{display:block;content:'\00D7';color:var(--drawer-close-color);position:relative;font-family:sans-serif;font-size:2rem;line-height:1;text-align:center}[type="checkbox"].drawer+* .drawer-close:hover,[type="checkbox"].drawer+* .drawer-close:focus{background:var(--drawer-hover-back-color)}@media screen and (max-width: 320px){[type="checkbox"].drawer+*{width:100%}}[type="checkbox"].drawer:checked+*{right:0}@media screen and (min-width: 768px){[type="checkbox"].drawer:not(.persistent)+*{position:static;height:100%;z-index:1100}[type="checkbox"].drawer:not(.persistent)+* .drawer-close{display:none}}:root{--table-border-color:#aaa;--table-border-separator-color:#666;--table-head-back-color:#e6e6e6;--table-head-fore-color:#111;--table-body-back-color:#f8f8f8;--table-body-fore-color:#111;--table-body-alt-back-color:#eee}table{border-collapse:separate;border-spacing:0;margin:0;display:flex;flex:0 1 auto;flex-flow:row wrap;padding:var(--universal-padding);padding-top:0}table caption{font-size:1.5rem;margin:calc(2 * var(--universal-margin)) 0;max-width:100%;flex:0 0 100%}table thead,table tbody{display:flex;flex-flow:row wrap;border:.0625rem solid var(--table-border-color)}table thead{z-index:999;border-radius:var(--universal-border-radius) var(--universal-border-radius) 0 0;border-bottom:.0625rem solid var(--table-border-separator-color)}table tbody{border-top:0;margin-top:calc(0 - var(--universal-margin));border-radius:0 0 var(--universal-border-radius) var(--universal-border-radius)}table tr{display:flex;padding:0}table th,table td{padding:calc(2 * var(--universal-padding))}table th{text-align:left;background:var(--table-head-back-color);color:var(--table-head-fore-color)}table td{background:var(--table-body-back-color);color:var(--table-body-fore-color);border-top:.0625rem solid var(--table-border-color)}table:not(.horizontal){overflow:auto;max-height:400px}table:not(.horizontal) thead,table:not(.horizontal) tbody{max-width:100%;flex:0 0 100%}table:not(.horizontal) tr{flex-flow:row wrap;flex:0 0 100%}table:not(.horizontal) th,table:not(.horizontal) td{flex:1 0 0%;overflow:hidden;text-overflow:ellipsis}table:not(.horizontal) thead{position:sticky;top:0}table:not(.horizontal) tbody tr:first-child td{border-top:0}table.horizontal{border:0}table.horizontal thead,table.horizontal tbody{border:0;flex:.2 0 0;flex-flow:row nowrap}table.horizontal tbody{overflow:auto;justify-content:space-between;flex:.8 0 0;margin-left:0;padding-bottom:calc(var(--universal-padding) / 4)}table.horizontal tr{flex-direction:column;flex:1 0 auto}table.horizontal th,table.horizontal td{width:auto;border:0;border-bottom:.0625rem solid var(--table-border-color)}table.horizontal th:not(:first-child),table.horizontal td:not(:first-child){border-top:0}table.horizontal th{text-align:right;border-left:.0625rem solid var(--table-border-color);border-right:.0625rem solid var(--table-border-separator-color)}table.horizontal thead tr:first-child{padding-left:0}table.horizontal th:first-child,table.horizontal td:first-child{border-top:.0625rem solid var(--table-border-color)}table.horizontal tbody tr:last-child td{border-right:.0625rem solid var(--table-border-color)}table.horizontal tbody tr:last-child td:first-child{border-top-right-radius:0.25rem}table.horizontal tbody tr:last-child td:last-child{border-bottom-right-radius:0.25rem}table.horizontal thead tr:first-child th:first-child{border-top-left-radius:0.25rem}table.horizontal thead tr:first-child th:last-child{border-bottom-left-radius:0.25rem}@media screen and (max-width: 767px){table,table.horizontal{border-collapse:collapse;border:0;width:100%;display:table}table thead,table th,table.horizontal thead,table.horizontal th{border:0;height:1px;width:1px;margin:-1px;overflow:hidden;padding:0;position:absolute;clip:rect(0 0 0 0);-webkit-clip-path:inset(100%);clip-path:inset(100%)}table tbody,table.horizontal tbody{border:0;display:table-row-group}table tr,table.horizontal tr{display:block;border:.0625rem solid var(--table-border-color);border-radius:var(--universal-border-radius);background:#fafafa;padding:var(--universal-padding);margin:var(--universal-margin);margin-bottom:calc(2 * var(--universal-margin))}table th,table td,table.horizontal th,table.horizontal td{width:auto}table td,table.horizontal td{display:block;border:0;text-align:right}table td:before,table.horizontal td:before{content:attr(data-label);float:left;font-weight:600}table th:first-child,table td:first-child,table.horizontal th:first-child,table.horizontal td:first-child{border-top:0}table tbody tr:last-child td,table.horizontal tbody tr:last-child td{border-right:0}}:root{--table-body-alt-back-color:#eee}table.striped tr:nth-of-type(2n)>td{background:var(--table-body-alt-back-color)}@media screen and (max-width: 768px){table.striped tr:nth-of-type(2n){background:var(--table-body-alt-back-color)}}:root{--table-body-hover-back-color:#90caf9}table.hoverable tr:hover,table.hoverable tr:hover>td,table.hoverable tr:focus,table.hoverable tr:focus>td{background:var(--table-body-hover-back-color)}@media screen and (max-width: 768px){table.hoverable tr:hover,table.hoverable tr:hover>td,table.hoverable tr:focus,table.hoverable tr:focus>td{background:var(--table-body-hover-back-color)}}:root{--mark-back-color:#0277bd;--mark-fore-color:#fafafa}mark{background:var(--mark-back-color);color:var(--mark-fore-color);font-size:.95em;line-height:1em;border-radius:var(--universal-border-radius);padding:calc(var(--universal-padding) / 4) calc(var(--universal-padding) / 2)}mark.inline-block{display:inline-block;font-size:1em;line-height:1.5;padding:calc(var(--universal-padding) / 2) var(--universal-padding)}:root{--toast-back-color:#424242;--toast-fore-color:#fafafa}.toast{position:fixed;bottom:calc(var(--universal-margin) * 3);left:50%;transform:translate(-50%, -50%);z-index:1111;color:var(--toast-fore-color);background:var(--toast-back-color);border-radius:calc(var(--universal-border-radius) * 16);padding:var(--universal-padding) calc(var(--universal-padding) * 3)}:root{--tooltip-back-color:#212121;--tooltip-fore-color:#fafafa}.tooltip{position:relative;display:inline-block}.tooltip:before,.tooltip:after{position:absolute;opacity:0;clip:rect(0 0 0 0);-webkit-clip-path:inset(100%);clip-path:inset(100%);transition:all 0.3s;z-index:1010;left:50%}.tooltip:not(.bottom):before,.tooltip:not(.bottom):after{bottom:75%}.tooltip.bottom:before,.tooltip.bottom:after{top:75%}.tooltip:hover:before,.tooltip:hover:after,.tooltip:focus:before,.tooltip:focus:after{opacity:1;clip:auto;-webkit-clip-path:inset(0%);clip-path:inset(0%)}.tooltip:before{content:'';background:transparent;border:var(--universal-margin) solid transparent;left:calc(50% - var(--universal-margin))}.tooltip:not(.bottom):before{border-top-color:#212121}.tooltip.bottom:before{border-bottom-color:#212121}.tooltip:after{content:attr(aria-label);color:var(--tooltip-fore-color);background:var(--tooltip-back-color);border-radius:var(--universal-border-radius);padding:var(--universal-padding);white-space:nowrap;transform:translateX(-50%)}.tooltip:not(.bottom):after{margin-bottom:calc(2 * var(--universal-margin))}.tooltip.bottom:after{margin-top:calc(2 * var(--universal-margin))}:root{--modal-overlay-color:rgba(0,0,0,0.45);--modal-close-color:#444;--modal-close-hover-color:#f0f0f0}[type="checkbox"].modal{height:1px;width:1px;margin:-1px;overflow:hidden;position:absolute;clip:rect(0 0 0 0);-webkit-clip-path:inset(100%);clip-path:inset(100%)}[type="checkbox"].modal+div{position:fixed;top:0;left:0;display:none;width:100vw;height:100vh;background:var(--modal-overlay-color)}[type="checkbox"].modal+div .card{margin:0 auto;max-height:50vh;overflow:auto}[type="checkbox"].modal+div .card .modal-close{position:absolute;top:0;right:0;width:1.75rem;height:1.75rem;border-radius:var(--universal-border-radius);padding:var(--universal-padding);margin:0;cursor:pointer;transition:background 0.3s}[type="checkbox"].modal+div .card .modal-close:before{display:block;content:'\00D7';color:var(--modal-close-color);position:relative;font-family:sans-serif;font-size:1.75rem;line-height:1;text-align:center}[type="checkbox"].modal+div .card .modal-close:hover,[type="checkbox"].modal+div .card .modal-close:focus{background:var(--modal-close-hover-color)}[type="checkbox"].modal:checked+div{display:flex;flex:0 1 auto;z-index:1200}[type="checkbox"].modal:checked+div .card .modal-close{z-index:1211}:root{--collapse-label-back-color:#e8e8e8;--collapse-label-fore-color:#212121;--collapse-label-hover-back-color:#f0f0f0;--collapse-selected-label-back-color:#ececec;--collapse-border-color:#ddd;--collapse-content-back-color:#fafafa;--collapse-selected-label-border-color:#0277bd}.collapse{width:calc(100% - 2 * var(--universal-margin));opacity:1;display:flex;flex-direction:column;margin:var(--universal-margin);border-radius:var(--universal-border-radius)}.collapse>[type="radio"],.collapse>[type="checkbox"]{height:1px;width:1px;margin:-1px;overflow:hidden;position:absolute;clip:rect(0 0 0 0);-webkit-clip-path:inset(100%);clip-path:inset(100%)}.collapse>label{flex-grow:1;display:inline-block;height:1.5rem;cursor:pointer;transition:background 0.3s;color:var(--collapse-label-fore-color);background:var(--collapse-label-back-color);border:.0625rem solid var(--collapse-border-color);padding:calc(1.5 * var(--universal-padding))}.collapse>label:hover,.collapse>label:focus{background:var(--collapse-label-hover-back-color)}.collapse>label+div{flex-basis:auto;height:1px;width:1px;margin:-1px;overflow:hidden;position:absolute;clip:rect(0 0 0 0);-webkit-clip-path:inset(100%);clip-path:inset(100%);transition:max-height 0.3s;max-height:1px}.collapse>:checked+label{background:var(--collapse-selected-label-back-color);border-bottom-color:var(--collapse-selected-label-border-color)}.collapse>:checked+label+div{box-sizing:border-box;position:relative;width:100%;height:auto;overflow:auto;margin:0;background:var(--collapse-content-back-color);border:.0625rem solid var(--collapse-border-color);border-top:0;padding:var(--universal-padding);clip:auto;-webkit-clip-path:inset(0%);clip-path:inset(0%);max-height:400px}.collapse>label:not(:first-of-type){border-top:0}.collapse>label:first-of-type{border-radius:var(--universal-border-radius) var(--universal-border-radius) 0 0}.collapse>label:last-of-type:not(:first-of-type){border-radius:0 0 var(--universal-border-radius) var(--universal-border-radius)}.collapse>label:last-of-type:first-of-type{border-radius:var(--universal-border-radius)}.collapse>:checked:last-of-type:not(:first-of-type)+label{border-radius:0}.collapse>:checked:last-of-type+label+div{border-radius:0 0 var(--universal-border-radius) var(--universal-border-radius)}mark.secondary{--mark-back-color:#d32f2f}mark.tertiary{--mark-back-color:#308732}mark.tag{padding:calc(var(--universal-padding)/2) var(--universal-padding);border-radius:1em}:root{--progress-back-color:#ddd;--progress-fore-color:#555}progress{display:block;vertical-align:baseline;-webkit-appearance:none;-moz-appearance:none;appearance:none;height:.75rem;width:calc(100% - 2 * var(--universal-margin));margin:var(--universal-margin);border:0;border-radius:calc(2 * var(--universal-border-radius));background:var(--progress-back-color);color:var(--progress-fore-color)}progress::-webkit-progress-value{background:var(--progress-fore-color);border-top-left-radius:calc(2 * var(--universal-border-radius));border-bottom-left-radius:calc(2 * var(--universal-border-radius))}progress::-webkit-progress-bar{background:var(--progress-back-color)}progress::-moz-progress-bar{background:var(--progress-fore-color);border-top-left-radius:calc(2 * var(--universal-border-radius));border-bottom-left-radius:calc(2 * var(--universal-border-radius))}progress[value="1000"]::-webkit-progress-value{border-radius:calc(2 * var(--universal-border-radius))}progress[value="1000"]::-moz-progress-bar{border-radius:calc(2 * var(--universal-border-radius))}progress.inline{display:inline-block;vertical-align:middle;width:60%}:root{--spinner-back-color:#ddd;--spinner-fore-color:#555}@keyframes spinner-donut-anim{0%{transform:rotate(0deg)}100%{transform:rotate(360deg)}}.spinner{display:inline-block;margin:var(--universal-margin);border:.25rem solid var(--spinner-back-color);border-left:.25rem solid var(--spinner-fore-color);border-radius:50%;width:1.25rem;height:1.25rem;animation:spinner-donut-anim 1.2s linear infinite}progress.primary{--progress-fore-color:#1976d2}progress.secondary{--progress-fore-color:#d32f2f}progress.tertiary{--progress-fore-color:#308732}.spinner.primary{--spinner-fore-color:#1976d2}.spinner.secondary{--spinner-fore-color:#d32f2f}.spinner.tertiary{--spinner-fore-color:#308732}span[class^='icon-']{display:inline-block;height:1em;width:1em;vertical-align:-0.125em;background-size:contain;margin:0 calc(var(--universal-margin) / 4)}span[class^='icon-'].secondary{-webkit-filter:invert(25%);filter:invert(25%)}span[class^='icon-'].inverse{-webkit-filter:invert(100%);filter:invert(100%)}span.icon-alert{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Ccircle cx='12' cy='12' r='10'%3E%3C/circle%3E%3Cline x1='12' y1='8' x2='12' y2='12'%3E%3C/line%3E%3Cline x1='12' y1='16' x2='12' y2='16'%3E%3C/line%3E%3C/svg%3E")}span.icon-bookmark{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='M19 21l-7-5-7 5V5a2 2 0 0 1 2-2h10a2 2 0 0 1 2 2z'%3E%3C/path%3E%3C/svg%3E")}span.icon-calendar{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Crect x='3' y='4' width='18' height='18' rx='2' ry='2'%3E%3C/rect%3E%3Cline x1='16' y1='2' x2='16' y2='6'%3E%3C/line%3E%3Cline x1='8' y1='2' x2='8' y2='6'%3E%3C/line%3E%3Cline x1='3' y1='10' x2='21' y2='10'%3E%3C/line%3E%3C/svg%3E")}span.icon-credit{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Crect x='1' y='4' width='22' height='16' rx='2' ry='2'%3E%3C/rect%3E%3Cline x1='1' y1='10' x2='23' y2='10'%3E%3C/line%3E%3C/svg%3E")}span.icon-edit{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='M20 14.66V20a2 2 0 0 1-2 2H4a2 2 0 0 1-2-2V6a2 2 0 0 1 2-2h5.34'%3E%3C/path%3E%3Cpolygon points='18 2 22 6 12 16 8 16 8 12 18 2'%3E%3C/polygon%3E%3C/svg%3E")}span.icon-link{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6'%3E%3C/path%3E%3Cpolyline points='15 3 21 3 21 9'%3E%3C/polyline%3E%3Cline x1='10' y1='14' x2='21' y2='3'%3E%3C/line%3E%3C/svg%3E")}span.icon-help{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3'%3E%3C/path%3E%3Ccircle cx='12' cy='12' r='10'%3E%3C/circle%3E%3Cline x1='12' y1='17' x2='12' y2='17'%3E%3C/line%3E%3C/svg%3E")}span.icon-home{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='M3 9l9-7 9 7v11a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2z'%3E%3C/path%3E%3Cpolyline points='9 22 9 12 15 12 15 22'%3E%3C/polyline%3E%3C/svg%3E")}span.icon-info{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Ccircle cx='12' cy='12' r='10'%3E%3C/circle%3E%3Cline x1='12' y1='16' x2='12' y2='12'%3E%3C/line%3E%3Cline x1='12' y1='8' x2='12' y2='8'%3E%3C/line%3E%3C/svg%3E")}span.icon-lock{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Crect x='3' y='11' width='18' height='11' rx='2' ry='2'%3E%3C/rect%3E%3Cpath d='M7 11V7a5 5 0 0 1 10 0v4'%3E%3C/path%3E%3C/svg%3E")}span.icon-mail{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='M4 4h16c1.1 0 2 .9 2 2v12c0 1.1-.9 2-2 2H4c-1.1 0-2-.9-2-2V6c0-1.1.9-2 2-2z'%3E%3C/path%3E%3Cpolyline points='22,6 12,13 2,6'%3E%3C/polyline%3E%3C/svg%3E")}span.icon-location{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='M21 10c0 7-9 13-9 13s-9-6-9-13a9 9 0 0 1 18 0z'%3E%3C/path%3E%3Ccircle cx='12' cy='10' r='3'%3E%3C/circle%3E%3C/svg%3E")}span.icon-phone{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='M22 16.92v3a2 2 0 0 1-2.18 2 19.79 19.79 0 0 1-8.63-3.07 19.5 19.5 0 0 1-6-6 19.79 19.79 0 0 1-3.07-8.67A2 2 0 0 1 4.11 2h3a2 2 0 0 1 2 1.72 12.84 12.84 0 0 0 .7 2.81 2 2 0 0 1-.45 2.11L8.09 9.91a16 16 0 0 0 6 6l1.27-1.27a2 2 0 0 1 2.11-.45 12.84 12.84 0 0 0 2.81.7A2 2 0 0 1 22 16.92z'%3E%3C/path%3E%3C/svg%3E")}span.icon-rss{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='M4 11a9 9 0 0 1 9 9'%3E%3C/path%3E%3Cpath d='M4 4a16 16 0 0 1 16 16'%3E%3C/path%3E%3Ccircle cx='5' cy='19' r='1'%3E%3C/circle%3E%3C/svg%3E")}span.icon-search{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Ccircle cx='11' cy='11' r='8'%3E%3C/circle%3E%3Cline x1='21' y1='21' x2='16.65' y2='16.65'%3E%3C/line%3E%3C/svg%3E")}span.icon-settings{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Ccircle cx='12' cy='12' r='3'%3E%3C/circle%3E%3Cpath d='M19.4 15a1.65 1.65 0 0 0 .33 1.82l.06.06a2 2 0 0 1 0 2.83 2 2 0 0 1-2.83 0l-.06-.06a1.65 1.65 0 0 0-1.82-.33 1.65 1.65 0 0 0-1 1.51V21a2 2 0 0 1-2 2 2 2 0 0 1-2-2v-.09A1.65 1.65 0 0 0 9 19.4a1.65 1.65 0 0 0-1.82.33l-.06.06a2 2 0 0 1-2.83 0 2 2 0 0 1 0-2.83l.06-.06a1.65 1.65 0 0 0 .33-1.82 1.65 1.65 0 0 0-1.51-1H3a2 2 0 0 1-2-2 2 2 0 0 1 2-2h.09A1.65 1.65 0 0 0 4.6 9a1.65 1.65 0 0 0-.33-1.82l-.06-.06a2 2 0 0 1 0-2.83 2 2 0 0 1 2.83 0l.06.06a1.65 1.65 0 0 0 1.82.33H9a1.65 1.65 0 0 0 1-1.51V3a2 2 0 0 1 2-2 2 2 0 0 1 2 2v.09a1.65 1.65 0 0 0 1 1.51 1.65 1.65 0 0 0 1.82-.33l.06-.06a2 2 0 0 1 2.83 0 2 2 0 0 1 0 2.83l-.06.06a1.65 1.65 0 0 0-.33 1.82V9a1.65 1.65 0 0 0 1.51 1H21a2 2 0 0 1 2 2 2 2 0 0 1-2 2h-.09a1.65 1.65 0 0 0-1.51 1z'%3E%3C/path%3E%3C/svg%3E")}span.icon-share{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Ccircle cx='18' cy='5' r='3'%3E%3C/circle%3E%3Ccircle cx='6' cy='12' r='3'%3E%3C/circle%3E%3Ccircle cx='18' cy='19' r='3'%3E%3C/circle%3E%3Cline x1='8.59' y1='13.51' x2='15.42' y2='17.49'%3E%3C/line%3E%3Cline x1='15.41' y1='6.51' x2='8.59' y2='10.49'%3E%3C/line%3E%3C/svg%3E")}span.icon-cart{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Ccircle cx='9' cy='21' r='1'%3E%3C/circle%3E%3Ccircle cx='20' cy='21' r='1'%3E%3C/circle%3E%3Cpath d='M1 1h4l2.68 13.39a2 2 0 0 0 2 1.61h9.72a2 2 0 0 0 2-1.61L23 6H6'%3E%3C/path%3E%3C/svg%3E")}span.icon-upload{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4'%3E%3C/path%3E%3Cpolyline points='17 8 12 3 7 8'%3E%3C/polyline%3E%3Cline x1='12' y1='3' x2='12' y2='15'%3E%3C/line%3E%3C/svg%3E")}span.icon-user{background-image:url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24' fill='none' stroke='%23111' stroke-width='2' stroke-linecap='round' stroke-linejoin='round'%3E%3Cpath d='M20 21v-2a4 4 0 0 0-4-4H8a4 4 0 0 0-4 4v2'%3E%3C/path%3E%3Ccircle cx='12' cy='7' r='4'%3E%3C/circle%3E%3C/svg%3E")}:root{--generic-border-color:rgba(0,0,0,0.3);--generic-box-shadow:0 .25rem .25rem 0 rgba(0,0,0,0.125),0 .125rem .125rem -.125rem rgba(0,0,0,0.25)}.hidden{display:none !important}.visually-hidden{position:absolute !important;width:1px !important;height:1px !important;margin:-1px !important;border:0 !important;padding:0 !important;clip:rect(0 0 0 0) !important;-webkit-clip-path:inset(100%) !important;clip-path:inset(100%) !important;overflow:hidden !important}.bordered{border:.0625rem solid var(--generic-border-color) !important}.rounded{border-radius:var(--universal-border-radius) !important}.circular{border-radius:50% !important}.shadowed{box-shadow:var(--generic-box-shadow) !important}.responsive-margin{margin:calc(var(--universal-margin) / 4) !important}@media screen and (min-width: 768px){.responsive-margin{margin:calc(var(--universal-margin) / 2) !important}}@media screen and (min-width: 1280px){.responsive-margin{margin:var(--universal-margin) !important}}.responsive-padding{padding:calc(var(--universal-padding) / 4) !important}@media screen and (min-width: 768px){.responsive-padding{padding:calc(var(--universal-padding) / 2) !important}}@media screen and (min-width: 1280px){.responsive-padding{padding:var(--universal-padding) !important}}@media screen and (max-width: 767px){.hidden-sm{display:none !important}}@media screen and (min-width: 768px) and (max-width: 1279px){.hidden-md{display:none !important}}@media screen and (min-width: 1280px){.hidden-lg{display:none !important}}@media screen and (max-width: 767px){.visually-hidden-sm{position:absolute !important;width:1px !important;height:1px !important;margin:-1px !important;border:0 !important;padding:0 !important;clip:rect(0 0 0 0) !important;-webkit-clip-path:inset(100%) !important;clip-path:inset(100%) !important;overflow:hidden !important}}@media screen and (min-width: 768px) and (max-width: 1279px){.visually-hidden-md{position:absolute !important;width:1px !important;height:1px !important;margin:-1px !important;border:0 !important;padding:0 !important;clip:rect(0 0 0 0) !important;-webkit-clip-path:inset(100%) !important;clip-path:inset(100%) !important;overflow:hidden !important}}@media screen and (min-width: 1280px){.visually-hidden-lg{position:absolute !important;width:1px !important;height:1px !important;margin:-1px !important;border:0 !important;padding:0 !important;clip:rect(0 0 0 0) !important;-webkit-clip-path:inset(100%) !important;clip-path:inset(100%) !important;overflow:hidden !important}}
</style>
</head>
<body>
<style type="text/css" charset="utf-8">
main {
  max-width: 48em;
  margin: auto;
}

@media only screen and (max-width: 768px) {
  main {
    max-width: 100%;
  }
}

img {
  max-width: 100%;
}
</style>


<main>
<header>
<h1 id="title" class="toc-ignore">Parus_major_RRBS_pipeline</h1>
<p id="date">5/25/2021</p>
<p id="author">Sonja Lecic</p>

</header>
<ul>
<li><a href="#bisulfate-conversion">Bisulfate Conversion</a>
<ul>
<li><a href="#file-preparation">File preparation</a>
<ul>
<li><a href="#chromosomes">Chromosomes</a></li>
<li><a href="#mitochondrion">Mitochondrion</a></li>
<li><a href="#repeats">Repeats</a></li>
<li><a href="#cut-sites">Cut-sites</a></li>
</ul></li>
<li><a href="#genome-preparation">Genome preparation</a>
<ul>
<li><a href="#libraries-list">Libraries list</a></li>
</ul></li>
<li><a href="#trimming-mapping-and-variant-calling">Trimming, Mapping and Variant calling</a>
<ul>
<li><a href="#the-full-pipeline">The full pipeline:</a></li>
</ul></li>
<li><a href="#cpg-count-summary">CpG count summary</a></li>
<li><a href="#methylation-bias-along-reads-m-bias">Methylation bias along reads (M-bias)</a></li>
<li><a href="#analysis-in-methylkit">Analysis in methylKit</a></li>
<li><a href="#statistical-analysis-with-lme4">Statistical analysis with lme4</a></li>
</ul></li>
<li><a href="#snp-calling">SNP calling</a>
<ul>
<li><a href="#merging-and-filtering">Merging and Filtering</a></li>
</ul></li>
<li><a href="#kinship-matrix">Kinship matrix</a></li>
</ul>

<style type="text/css">
.columns {display: flex;}
h1 {color: purple;}
h2 {color: darkorange;}
</style>
<h1 id="bisulfate-conversion">Bisulfate Conversion</h1>
<p>Parus major latest genome version (v1.1) from April 2020 can be found <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_001522545.3#/def_asm_Primary_Assembly">here</a>.</p>
<p>Cyanistes carules latest genome version (v1) from January 2018 can be found <a href="https://www.ncbi.nlm.nih.gov/assembly/GCF_002901205.1/">here</a>.</p>
<h2 id="file-preparation">File preparation</h2>
<h3 id="chromosomes">Chromosomes</h3>
<p>Extract chromosomes bed and genome files from *fa.fai. Chromosomes.bed file contains #chr #start #end. Chromosomes genome file contains #chr #FullLength.</p>
<pre><code># make a .fai file with samtools
samtools faidx Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa
# grep all the chromsomes and make a list
grep &#39;chr&#39; Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa.fai &gt; Chromosomes.list
#Convert 1-based index to 0-based .bed format 
awk &#39;{OFS=&quot;\t&quot;}{print $1, &quot;0&quot;, $2-1}&#39; Chromosomes.list &gt; Chromosomes.bed
#Create a genome length file for bedtools as well, called a &#39;genome&#39; file 
awk &#39;{OFS=&quot;\t&quot;}{print $1, $2-1}&#39; Chromosomes.list &gt; Chromosomes.genome</code></pre>
<h3 id="mitochondrion">Mitochondrion</h3>
<p>Create a .bed file from mitochondrion separately.</p>
<pre><code># grep the mitochondrion in a list
grep &#39;chrM&#39; Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa.fai &gt; mtDNA.list
#Convert 1-based index to 0-based .bed format 
awk &#39;{OFS=&quot;\t&quot;}{print $1, &quot;0&quot;, $2-1}&#39; mtDNA.list &gt; mtDNA.bed</code></pre>
<h3 id="repeats">Repeats</h3>
<p>Create a bed file containing repeats. These will be analyzed separately. To extract the repeat I will use RepeatMasker.</p>
<pre><code>#!/bin/bash

#SBATCH --get-user-env
#SBATCH --mail-user=lecic@bio.lmu.de
#SBATCH --clusters=biohpc_gen
#SBATCH --partition=biohpc_gen_normal
#SBATCH --cpus-per-task=20
#SBATCH --mem-per-cpu=4763mb
#SBATCH --time=12:00:00

#specify path to the genome and genome directory
genome=/dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1/Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa
genomedir=/dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1

RepeatMasker ${genome} -species chicken -nolow -pa 20 -gff -dir ${genomedir}</code></pre>
<p>Convert it to a standard bed:</p>
<pre><code>awk &#39;{print $1, $4, $5, $7, $10}&#39; Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa.out.gff &gt; repeats.tmp1
grep -E -- &#39;chr&#39; repeats.tmp1 &gt; repeats_chrom.tmp
awk &#39;{OFS=&quot;\t&quot;}{print $1, $2, $3, $4, $5}&#39; repeats_chrom.tmp | sed &#39;s/&quot;//g&#39; &gt; Parus_Repeats.bed</code></pre>
<h3 id="cut-sites">Cut-sites</h3>
<p>Identify cut-sites in the genome. For this, I will use a combination of an R and bash scripts.The script uses simRAD R package to find the cut-sites. I we are going to use the <em>MspI restriction enzyme</em> and <em>fragment size selection of 100-400bp</em> this needs to be put in the function. Cut-sites bed file will be used to after mapping of the RRBS reads to subset the reads that overlap with these fragments in order to reduce the number erroneously mapped reads.</p>
<p>But, first we need to prepare the bed file for the whole genome (chromosomes + scaffolds; we can ectract from the final file with cutsites whatever we want then). For this Iâ€™ll use bioawk to make a bed file (#chrom #start #end) in a 0 based bed format:</p>
<pre><code>bioawk -c fastx &#39;{print $name&quot;\t0\t&quot;length($seq)-1}&#39; Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa &gt; Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.bed</code></pre>
<p>And, now the R script:</p>
<pre><code>#!/usr/bin/env Rscript


# Author:Sonja Lecic
# Originally written by Justin MerÃ¶ndun
# Modified by Sonja on 20.03.2021 to have the whole pipeline run faster 

#######################################
## In silico digest with SimRAD ###
#######################################
# set the working directory to a folder where all the files will be created and saved
setwd(&quot;/Volumes/LaCie/PhD/data/genomes/BlueTit/&quot;)

# call bash script from the working directory (find explanations of what the script does directly in the script)
x &lt;- &#39;bash /Volumes/LaCie/PhD/data/RRBSParus/cluster/scripts/chrSelect.sh&#39;
system(x)

# load SimRAD library
library(SimRAD)
# make a function to create text a file with fragment ranges that takes as input: 
# - 5&#39; and 3&#39; nucleotide pattern that restriction enzyme recognizes
# - min and max size of selection of the fragments
restrictionEnzyme &lt;- function(cs_5p1, cs_3p1, minsize, maxsize) {
  # make a list of all SEQ.fasta files
  myFiles &lt;- list.files(pattern = &quot;\\.SEQ.fasta$&quot;) 
  for(i in myFiles){
    bluetit &lt;- ref.DNAseq(i, subselect.contigs=FALSE)
    # pattern that enzyme recognizes; for MspI : C&#39;CGG
    cs_5p1 = cs_5p1
    cs_3p1 = cs_3p1
    bluetit.dig &lt;- insilico.digest(bluetit, cs_5p1, cs_3p1, verbose=T)
    bluetit.sel &lt;- adapt.select(bluetit.dig, type=&quot;AA&quot;, cs_5p1, cs_3p1)
    nar.bluetit &lt;- size.select(bluetit.sel,  min.size = minsize, max.size = maxsize, graph=F, verbose=T)
    dt &lt;- as.data.frame(nar.bluetit@ranges)
    chr &lt;- rep(gsub(pattern = &quot;\\.SEQ.fasta$&quot;, &quot;&quot;, (i)), nrow(dt))
    dtf &lt;- cbind(dt, chr)
    bed &lt;- dtf[, c(5, 2 ,3)]
    outputFile &lt;- paste0(tools::file_path_sans_ext(i), &quot;.bed&quot;)
    write.table(bed, file = outputFile, col.names = F, row.names = F, sep = &quot;\t&quot;, quote = FALSE)
  }
}

## call the function with desired restriction pattern and fragment size
restrictionEnzyme(cs_5p1 = &quot;C&quot;, cs_3p1 = &quot;CGG&quot;, minsize = 100, maxsize = 400)


## remove empty files if there are any (because some parts of a chromosome will have no cutsites and therefore produce emty bed files)
docs &lt;- list.files(pattern = &quot;\\.SEQ.bed$&quot;)   
inds &lt;- file.size(docs) == 0 
file.remove(docs[inds])

# concatenate all the beds into one bed file with #Chrom #Start #End as columns
myBeds &lt;- list.files(pattern = &quot;\\.SEQ.bed$&quot;)
cutsites &lt;-  read.table(myBeds[1], header = F)
for (f in myBeds[-1]){
  cutsites &lt;- rbind(cutsites, read.table(f))
}
# give a name to your file and save it in your working directory
write.table(cutsites, file = &quot;CutsitesParusMspI100_400.bed&quot;, col.names = F, row.names = F, sep = &quot;\t&quot;, quote = FALSE)</code></pre>
<p>..and the bash script <em>chrSelect.sh</em> that is called within the R script:</p>
<pre><code>#!/bin/bash
### this script is selecting each scaffold from the genome fasta and saving it as a separate fasta file

for i in $(awk &#39;{print $1}&#39; /Volumes/LaCie/genome_assemblies_genome_fasta_Parus/ncbi-genomes-2021-05-20/Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.bed); do
#simRAD can only hold on ~250 mb scaffolds at once, subselect scaffold
grep -w ${i} /Volumes/LaCie/genome_assemblies_genome_fasta_Parus/ncbi-genomes-2021-05-20/Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.bed &gt; ${i}.GRAB.bed
#use seqtk to grab this scaffold from the genome and save it as a separate file 
seqtk subseq /Volumes/LaCie/genome_assemblies_genome_fasta_Parus/ncbi-genomes-2021-05-20/Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa ${i}.GRAB.bed &gt; ${i}.SEQ.fasta
wait
# remove the GRAB.bed file from the folder
rm *.GRAB.bed
done</code></pre>
<h2 id="genome-preparation">Genome preparation</h2>
<p>Genome will be mapped with bismark. Before the actuall mapping the genome needs to be prepared:</p>
<pre><code>#!/bin/bash

#SBATCH --get-user-env
#SBATCH --mail-user=lecic@bio.lmu.de
#SBATCH --clusters=biohpc_gen
#SBATCH --partition=biohpc_gen_normal
#SBATCH --cpus-per-task=3
#SBATCH --mem-per-cpu=4763mb
#SBATCH --time=12:00:00

SCRATCH=/tmp/$SLURM_JOB_ID

#path to genome directory
genomedir=/dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1

bismark_genome_preparation ${genomedir}</code></pre>
<h3 id="libraries-list">Libraries list</h3>
<p>Make a list of libraries for Parus major because I will make a script that will take each library in parallel from the raw directory and run it through the whole mapping pipeline.</p>
<pre><code>ls *.fastq.gz | rev | cut -c10- | rev &gt; /dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1/Libraries.list</code></pre>
<p>And, the genome folder looks like this now:</p>
<pre><code>total 2.1G
drwxrws--- 4 ra57xaf pr53da-dss-0024 4.0K Jun  9 09:40 Bisulfite_Genome   # output of bismark genome preparation
-rw-rw---- 1 ra57xaf pr53da-dss-0024  557 May 26 13:35 Chromosomes.bed   # coordinates of major chromosomes
-rw-rw---- 1 ra57xaf pr53da-dss-0024  491 May 26 13:32 Chromosomes.genome   # total lengths of all major chromosomes
-rw-rw---- 1 ra57xaf pr53da-dss-0024 2.9M Jun  9 12:21 CutsitesParusMspI100_400.bed   # coodinates of the putative cutsites
-rw-rw---- 1 ra57xaf pr53da-dss-0024  816 Jun  9 11:43 Libraries.list   # list of Parus libraries
-rw-rw---- 1 ra57xaf pr53da-dss-0024   13 May 26 13:40 mtDNA.bed   # coodinates of mtDNA
-rw-rw---- 1 ra57xaf pr53da-dss-0024  14M Jun  9 13:08 Parus_C2Tsnps.bed   # coordinates of C-T SNPs
-rw-rw---- 1 ra57xaf pr53da-dss-0024 986M May 26 12:38 Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa # Parus genome fasta
-rw-rw---- 1 ra57xaf pr53da-dss-0024  34M Jun  9 11:35 Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa.cat.gz
-rw-rw---- 1 ra57xaf pr53da-dss-0024 993M Jun  9 11:35 Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa.masked
-rw-rw---- 1 ra57xaf pr53da-dss-0024 9.0M Jun  9 11:35 Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa.out
-rw-rw---- 1 ra57xaf pr53da-dss-0024 5.9M Jun  9 11:35 Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa.out.gff
-rw-rw---- 1 ra57xaf pr53da-dss-0024 2.5K Jun  9 11:35 Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa.tbl
-rw-rw---- 1 ra57xaf pr53da-dss-0024 2.7M Jun  9 11:52 Parus_Repeats.bed   # coodinates of the repeats
-rw-rw---- 1 ra57xaf pr53da-dss-0024 2.8M Jun  9 11:51 repeats_chrom.tmp
-rw-rw---- 1 ra57xaf pr53da-dss-0024 2.9M Jun  9 11:47 repeats.tmp1
drwxrws--- 2 ra57xaf pr53da-dss-0024 4.0K Jun  9 11:35 scripts_genome_prep   # folder with scripts
-rw-rw---- 1 ra57xaf pr53da-dss-0024    0 Jun  8 14:57 Snakefile</code></pre>
<h2 id="trimming-mapping-and-variant-calling">Trimming, Mapping and Variant calling</h2>
<h3 id="the-full-pipeline">The full pipeline:</h3>
<p>RRBS_Pipeline_CG.sh script</p>
<pre><code>#!/bin/bash

#SBATCH --get-user-env
#SBATCH --mail-user=lecic@bio.lmu.de
#SBATCH --clusters=biohpc_gen
#SBATCH --partition=biohpc_gen_normal
#SBATCH --cpus-per-task=10
#SBATCH --mem-per-cpu=4763mb
#SBATCH --time=24:00:00

# The Files
# genome folder
genomedir=/dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1
# Parus major genome fasta
genome=${genomedir}/Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa
# putative cutsites simulated with simRad
cutsites=${genomedir}/CutsitesParusMspI100_400.bed
# list of all SNPs
snps=${genomedir}/Parus_SNPs.bed
# bed file with coordinates of the repeats
repeats=${genomedir}/Parus_Repeats.bed
# bed file with lengths of all major chromosomes
chrs=${genomedir}/Chromosomes.genome
# bed file with coordinates of major chromosomes
chrsbed=${genomedir}/Chromosomes.bed
# bed file with coordinates of the mitochondrion
mtdna=${genomedir}/mtDNA.bed
#bed file with 0-based coordinates for C-T and A-G SNPs
c2tsnp=${genomedir}/Parus_C2Tsnps.bed

#Directory paths
rawdata=/dss/dsslegfs01/pr53da/pr53da-dss-0024/rawdata/Bisulfite-Seq/RRBS
workdir=/dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1/work
outdir=/dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1/out

mkdir ${workdir}
mkdir ${outdir}

RUN=$1

#Trim adapters (for single-end reads)
trim_galore --fastqc -j 8 --quality 30 --rrbs --output_dir ${workdir} ${rawdata}/${RUN}*.fastq.gz

#Map reads
bismark --parallel 10 --output_dir ${workdir} --genome ${genomedir} ${workdir}/${RUN}_trimmed.fq.gz

#Extract methylation
bismark_methylation_extractor --parallel 6 --gzip --bedgraph --buffer_size 20G --cytosine_report --genome_folder ${genomedir} --output ${workdir}/ ${workdir}/${RUN}_trimmed_bismark_bt2.bam

#Only keep positions within predicted inserts
bedtools intersect -wb -a ${cutsites} -b ${workdir}/${RUN}*_trimmed_bismark_bt2.bismark.cov.gz | awk &#39;{OFS=&quot;\t&quot;}{print $4, $5, $6, $7, $8, $9}&#39; | sort | uniq &gt; ${workdir}/${RUN}.filt_cutsite.tmp

#Remove any positions overlapping a transition (C-T SNPs)
bedtools subtract -A -a ${workdir}/${RUN}.filt_cutsite.tmp -b ${c2tsnp} &gt; ${workdir}/${RUN}.filt_cutsite_c2tsnp.tmp

#Only keep sites on the major chromosomes + and sort according to chromosome and position
bedtools intersect -wb -a ${chrsbed} -b ${workdir}/${RUN}.filt_cutsite_c2tsnp.tmp | awk &#39;{OFS=&quot;\t&quot;}{print $4, $5, $6, $7, $8, $9}&#39; | sort -k1,1 -k2,2n | uniq &gt; ${workdir}/${RUN}.filt_cutsite_c2tsnp_chrs.tmp

#Divide files further into repeat-region 5mC, and then our final 5mC positions for analysis (+ and sort according to chromosome and position)
bedtools intersect -wb -a ${repeats} -b ${workdir}/${RUN}.filt_cutsite_c2tsnp_chrs.tmp | awk &#39;{OFS=&quot;\t&quot;}{print $4, $5, $6, $7, $8, $9}&#39; | sort -k1,1 -k2,2n | uniq | bgzip -c &gt; ${outdir}/${RUN}.CpG_in_Repeats.cov.gz

#Subtract repeats for the FINAL file for our analysis
bedtools subtract -A -a ${workdir}/${RUN}.filt_cutsite_c2tsnp_chrs.tmp -b ${repeats} | bgzip -c &gt; ${outdir}/${RUN}.CpG_5mC.cov.gz

###########

### SUMMARIZE COUNTS
#Starting, raw positions
echo &quot;RAW&quot; &gt;&gt; ${outdir}/${RUN}.COUNTS.txt
zcat ${workdir}/${RUN}*_trimmed_bismark_bt2.bismark.cov.gz | wc -l &gt;&gt; ${outdir}/${RUN}.COUNTS.txt

#Positions remaining after cut-site filter
echo &quot;CUTSITE_FILTER&quot; &gt;&gt; ${outdir}/${RUN}.COUNTS.txt
cat ${workdir}/${RUN}.filt_cutsite.tmp | wc -l &gt;&gt; ${outdir}/${RUN}.COUNTS.txt

#Positions remaining after C-T SNP filter
echo &quot;CT_SNP_FILTER&quot; &gt;&gt; ${outdir}/${RUN}.COUNTS.txt
cat ${workdir}/${RUN}.filt_cutsite_c2tsnp.tmp | wc -l &gt;&gt; ${outdir}/${RUN}.COUNTS.txt

#Positions remaining on major chromosomes
echo &quot;CHROMOSOME_FILTER&quot; &gt;&gt; ${outdir}/${RUN}.COUNTS.txt
cat ${workdir}/${RUN}.filt_cutsite_c2tsnp_chrs.tmp | wc -l &gt;&gt; ${outdir}/${RUN}.COUNTS.txt

#Positions remaining after filtering sites within repeats
echo &quot;REPEAT_FILTER&quot; &gt;&gt; ${outdir}/${RUN}.COUNTS.txt
zcat ${outdir}/${RUN}.CpG_5mC.cov.gz | wc -l &gt;&gt; ${outdir}/${RUN}.COUNTS.txt

#Add a name column for each library for easier manipulation later
echo &quot;LIBRARY_NAME&quot; &gt;&gt; ${outdir}/${RUN}.COUNTS.txt
fbname=$(basename ${outdir}/${RUN}.COUNTS.txt | sed -E &#39;s/(([^_]*_){6}).*/\1/; s/_$//&#39; )
echo &quot;$fbname&quot; &gt;&gt; ${outdir}/${RUN}.COUNTS.txt

#Transpose the table so that we have 6 columns and one row with counts
cat ${outdir}/${RUN}.COUNTS.txt | pr -ts&#39; &#39; --column 6 &gt; ${outdir}/${RUN}.tCOUNTS.txt
</code></pre>
<p>And create a bash script for each sample:</p>
<pre><code>for sample in $(cat /dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1/Libraries.list); do sbatch -J ${sample} RRBS_Pipeline_CG.sh ${sample}; done </code></pre>
<p>et voila! The final file .CpG_5mC.cov.gz looks like this:</p>
<pre><code>The final file contains 6 columns: #Chromosome #Start #End #Methylation_Percentage #Count_Methylated #Count_Unmethylated

zcat C2R0242_BL_ADL_F__SRR10606832__SE__976ce27c8be6fb513c6153f49b03a3a1.CpG_5mC.cov.gz | head
chr01   4634    4634    88.4057971014493    183 24
chr01   4641    4641    88.4615384615385    184 24
chr01   4650    4650    82.6923076923077    172 36
chr01   4673    4673    97.1153846153846    202 6
chr01   4682    4682    72.7272727272727    8    3
chr01   4711    4711    100                 1    0
chr01   4803    4803    17.3913043478261    12  57
chr01   4817    4817    34.7826086956522    24  45
chr01   4860    4860    100                 1    0
chr01   4865    4865    50                  1    1</code></pre>
<h2 id="cpg-count-summary">CpG count summary</h2>
<p>Concatenate all COUNT files into one to make an R input file:</p>
<pre><code>awk &#39;FNR&gt;1 || NR==1&#39; *tCOUNTS* &gt; COUNTS.input</code></pre>
<p>And R code:</p>
<pre><code>library(ggplot2)
library(ggsci)

counts &lt;- read.table(&quot;/Volumes/LaCie/PhD/data/RRBSParus/cluster/COUNTS.input&quot;,header=T)
head(counts)

ggcounts &lt;- data.frame(Filter = c(rep(colnames(counts[1]), nrow(counts)), 
                                  rep(colnames(counts[2]), nrow(counts)),
                                  rep(colnames(counts[3]), nrow(counts)),
                                  rep(colnames(counts[4]), nrow(counts)),
                                  rep(colnames(counts[5]), nrow(counts))),
                       Counts = c(unlist(counts[, c(1:5)], use.names = FALSE)),
                       Library = c(paste(rep(counts$LIBRARY_NAME))))

head(ggcounts)

png(&quot;ParusCpGcounts.png&quot;, width = 30, height = 30, units = &quot;cm&quot;, res=600)
ggplot(ggcounts,aes(x= Library, fill= reorder(Filter,-Counts), y=Counts))+
  geom_bar(stat=&quot;identity&quot;,position=&quot;dodge&quot;)+
  theme_classic(base_size=20)+
  labs(fill=&#39;Filter&#39;) +
  #scale_fill_manual(values=mypal)+
  scale_color_futurama()+
  ylab(&quot;Number of CpGs&quot;)+
  theme(axis.text.x=(element_text(angle=65,hjust=1)))
dev.off()
</code></pre>
<h2 id="methylation-bias-along-reads-m-bias">Methylation bias along reads (M-bias)</h2>
<p>Within /work/ there will be many files with an m-bias output showing methylation biases along read lengths, which could indicate systematic biases in the data.</p>
<p>The file contains outputs for CpGs, CHG and CHH for each sample and looks like this:</p>
<pre><code>head B2X6537_BL_CHK_M__SRR10606831__SE__c0c278066ffc4182f5ab186575ff6d0d_trimmed_bismark_bt2.M-bias.txt
CpG context
===========
position    count methylated    count unmethylated  % methylation   coverage
1   5338097 8086304 39.76   13424401
2   634 812 43.85   1446
3   396 1030    27.77   1426
4   108026  335702  24.35   443728
5   123684  282038  30.48   405722
6   166299  425252  28.11   591551
7   176873  468122  27.42   644995</code></pre>
<p>Combine all the outputs only for CpGs. This command will merge all the files taking only CpGs, and remove the first 3 lines.</p>
<pre><code>grep -A 53 &#39;CpG&#39; *M-bias* | sed &#39;/CpG/d&#39; | sed &#39;/======/d&#39; | sed &#39;/--/d&#39; | sed &#39;/count/d&#39; &gt; ../out/M_BIAS_CpG.input</code></pre>
<p>And R code:</p>
<pre><code>mbias &lt;- read.table(&quot;/Volumes/LaCie/PhD/data/RRBSParus/cluster/download_files/M_BIAS_CpG.input&quot;)

mbias$position &lt;- paste0(substr(mbias$V1, 100,101))
mbias$V1 &lt;- paste0(substr(mbias$V1, 1,29))
Mbiasnames &lt;- c(&quot;Sample&quot;, &quot;Count_M&quot;, &quot;Count_U&quot;, &quot;Percent_M&quot;, &quot;Coverage&quot;, &quot;Position&quot;)
names(mbias) &lt;- Mbiasnames

# make a neat function to reorder columns (for later analysis also)
moveCol &lt;- function(data, col2move, where = &quot;last&quot;, beforeafter = NULL) {
  temp &lt;- setdiff(names(data), col2move)
  x &lt;- switch(
    where,
    first = data[c(col2move, temp)],
    last = data[c(temp, col2move)],
    before = {
      if (is.null(beforeafter)) stop(&quot;must specify beforeafter column&quot;)
      if (length(beforeafter) &gt; 1) stop(&quot;beforeafter must be a single character string&quot;)
      data[append(temp, values = col2move, after = (match(beforeafter, temp)-1))]
    },
    after = {
      if (is.null(beforeafter)) stop(&quot;must specify beforeafter column&quot;)
      if (length(beforeafter) &gt; 1) stop(&quot;beforeafter must be a single character string&quot;)
      data[append(temp, values = col2move, after = (match(beforeafter, temp)))]
    })
  x
}
mbiasdf = moveCol(data = mbias, col2move = &quot;Position&quot;, where = &quot;after&quot;, beforeafter = &quot;Sample&quot;)

# change positions column to numeric
mbiasdf$Position &lt;- as.numeric(mbiasdf$Position)

# plot M-bias
library(ggsci)
library(ggplot2)

ggplot(mbiasdf,aes(x=Position, col=Sample))+
  geom_line(aes(y=Percent_M),stat=&quot;identity&quot;,show.legend=FALSE, group = 12)+
  theme_classic(base_size=16)+
  scale_color_futurama()+
  coord_cartesian(ylim=c(0,100))

ggplot(mbiasdf,aes(x=Position,col=Sample))+
  geom_line(aes(y=Coverage),stat=&quot;identity&quot;,show.legend=FALSE, group = 12)+
  theme_classic(base_size=16)+
  scale_color_futurama()
</code></pre>
<h2 id="analysis-in-methylkit">Analysis in methylKit</h2>
<p>And R code:</p>
<pre><code>library(methylKit)

# create a list of all bismark cov files
file.list=list.files(path = &quot;/Volumes/LaCie/PhD/data/RRBSParus/cluster/download_files&quot;, pattern = &#39;[gz]&#39;, full.names = F)
file.list=as.list(file.list)

# create a list of sample IDs from file names
sampleID=list()
for (i in 1:length(file.list)) {
  name = substring(file.list[[i]], 1, 29)
  print(name)
  sampleID=append(sampleID, name)
}

# read files in methylKit
myobj=methRead(file.list,
               pipeline = &quot;bismarkCoverage&quot;,
               sample.id=sampleID,
               assembly = &quot;Parus.major.v1.1&quot;,
               treatment=c(rep(1, 12)))
               

# filter for coverage above 20
filtered.myobj=filterByCoverage(myobj,lo.count=20,lo.perc=NULL,
                                hi.count=NULL,hi.perc=99.9)
                                
# normalize coverage
filt.norm.myobj=normalizeCoverage(filtered.myobj, method=&quot;median&quot;)

# unite all samples into one table
meth=methylKit::unite(filt.norm.myobj, destrand=FALSE, min.per.group = 9L)
head(meth)
dim(meth)</code></pre>
<h4 id="diffetentially-methylted-sites">Diffetentially methylted sites</h4>
<p>And R code:</p>
<pre><code># Calculate differentially methylated sites (not regions) using Fisher&#39;s exact test and BH correction
myDiff=calculateDiffMeth(meth, adjust = &quot;BH&quot;, test = &quot;fast.fisher&quot;, slim = F)
# BH cutoff
myDiff001=getMethylDiff(myDiff,difference=25,qvalue=0.01)
cutoff &lt;- -log10(myDiff001$qvalue)

data &lt;- data.frame(chr = myDiff$chr,
                   win = myDiff$start,
                   pval= -log10(myDiff$qvalue))

chrlen &lt;- read.table(&quot;Chromosomes.genome&quot;)
chrlen=chrlen[order(chrlen$V1),]
chromosomes=unique(myDiff$chr)
chromosomes_length=matrix(chrlen$V2, nrow = 1, dimnames = list(NULL, chromosomes))
off &lt;- 0
pval_lim &lt;- max(as.numeric(-log10(myDiff$qvalue)))
png(&quot;libraryName.png&quot;, width = 3000, height = 1200, units = &quot;px&quot;, pointsize = 12)
for(i in seq_along(chromosomes)){
  chr &lt;- chromosomes[i]
  if(i%%2){col &lt;- &quot;deepskyblue4&quot;}else{col &lt;- &quot;darkorange1&quot;}
  temp &lt;- data[which(data$chr == chr), ]
  print(head(temp))
  if(chr == chromosomes[1]){
    plot(temp$win, temp$pval, xaxt = &quot;n&quot;, xlim = c(0, sum(chromosomes_length[1, ])), ylim = c(0, pval_lim), col = col, pch=19,
         ylab = &quot;-log10(p)&quot;, xlab = &quot;chromosomes&quot;, cex.lab=1.5)
  }else{
    points(temp$win+off, temp$pval, col = col, pch=20)
  }
  off &lt;- off + chromosomes_length[1, chr]
  abline(h=min(cutoff), col = &quot;red&quot;)
  chrl &lt;- c(0, chrlen$V2)
  v=c(0 + cumsum(chrl))
  axis(1, at=v[-length(v)] + diff(v) / 2, labels=chromosomes, las=2)
}
dev.off()
</code></pre>
<h2 id="statistical-analysis-with-lme4">Statistical analysis with lme4</h2>
<pre><code>library(lme4)
library(lmerTest)

### pepare methylation data frame for statistical analysis
head(meth)
dim(meth)

methdf &lt;- methylKit::getData(meth)
samplelist &lt;- meth@sample.ids

# make a function to get a table that combines methylation calls and trios information for statistical analysis
makeMethData &lt;- function(methdf, samplelist, trios) {
  
  sampleName=vector()
  for (i in 1:length(samplelist)) {
    name = substring(samplelist[i], 1, 29)
    sampleName=append(sampleName, name)
  }
  
  logitTransform &lt;- function(p) { log(p/(1-p)) }
  n &lt;- 3
  cov &lt;- methdf[5:ncol(methdf)][, c(TRUE, rep(FALSE, n - 1))]
  numC &lt;- methdf[5:ncol(methdf)][, c(FALSE, TRUE, FALSE)]
  
  fC &lt;- cbind(methdf[1:2],logitTransform(numC/cov))
  
  sampleName=as.character(sampleName)
  names(fC)[-c(1:2)] &lt;- sampleName
  
  print(fC)
  
  # trios is a data set with trios ID and sex, petrnity, brood, nestbox, sampling year information
  # put everything together in a final data table
  epidata = data.frame(chr = c(rep(fC$chr, length(trios$offspring))),
                        pos = c(rep(fC$start, length(trios$offspring))),
                        fCoffspring = unlist(fC[, trios$offspring], use.names = FALSE),
                        fCmom = unlist(fC[, trios$mom], use.names = FALSE),
                        fCdad = unlist(fC[, trios$dad], use.names = FALSE),
                        sex = c(rep(trios$sex, each=nrow(fC))),
                        lifestage = c(rep(trios$lifestage, each=nrow(fC))),
                        paternity = c(rep(trios$paternity, each=nrow(fC))),
                        brood = c(rep(trios$brood, each=nrow(fC))),
                        nestbox = c(rep(trios$nestbox, each=nrow(fC))),
                        sampyear = c(rep(trios$sampyear, each=nrow(fC))))
  print(epidata)
  
}
# call the function
epistats = makeMethData(methdf = methdf, samplelist = samplelist, trios = trios)

# convert any Infinity or NaN values to NA 
epistats[epistats == &quot;-Inf&quot;] &lt;- NA
epistats[epistats == &quot;Inf&quot;] &lt;- NA
epistats[epistats == &quot;NaN&quot;] &lt;- NA

# call an lmm model
epi.lmer &lt;- lmer(fCoffspring ~  fCmom + (1|lifestage), data = epistats)
episum &lt;- summary(epi.lmer)
# plot
plot(epi.lmer)
qqnorm(resid(epi.lmer))

# make a nice table out of summary results with stargazer library
library(stargazer)
# change the class to make it compatible for stargazer
class(epi.lmer) &lt;- &quot;lmerMod&quot;
epi.lmertable &lt;- stargazer(epi.lmer, type = &quot;text&quot;,
          digits = 3,
          star.cutoffs = c(0.05, 0.01, 0.001),
          digit.separator = &quot;&quot;)

</code></pre>
<h1 id="snp-calling">SNP calling</h1>
<p>Call SNPs from the RRBS mapped files in order to get all SNPs (A, T, C and G). Frist, I will sort the bam files and add read groups with picard. As there is no need to realign around indels (Bismark/Bowtie2 allows for gapped aligments). Then, base quality recalibration needs to be done as Bis-SNP heavily depends on base quality score for the genotype probability calculation, but Illumina sequencing readsâ€™ raw base quality score could not accurately reflect the error rate of the base. I will then call SNPs with bis-snp. RRBS_SNPcall.sh script:</p>
<pre><code>#!/bin/bash

#SBATCH --get-user-env
#SBATCH --mail-user=lecic@bio.lmu.de
#SBATCH --clusters=biohpc_gen
#SBATCH --partition=biohpc_gen_normal
#SBATCH --cpus-per-task=10
#SBATCH --mem-per-cpu=4763mb
#SBATCH --time=24:00:00

# The Files
# genome folder
genomedir=/dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1
# Parus major genome fasta
genome=${genomedir}/Parus.major_genome_Abel_Illumina.HiSeq_AllPathsv.April2013_v1.1.fa
# putative cutsites simulated with simRad
cutsites=${genomedir}/CutsitesParusMspI100_400.bed
# list of all SNPs
snps=${genomedir}/Parus_SNPs.bed
# bed file with coordinates of the repeats
repeats=${genomedir}/Parus_Repeats.bed
# bed file with lengths of all major chromosomes
chrs=${genomedir}/Chromosomes.genome
# bed file with coordinates of major chromosomes
chrsbed=${genomedir}/Chromosomes.bed
# bed file with coordinates of the mitochondrion
mtdna=${genomedir}/mtDNA.bed
#bed file with 0-based coordinates for C-T and A-G SNPs
c2tsnp=${genomedir}/Parus_C2Tsnps.bed

#Directory paths
rawdata=/dss/dsslegfs01/pr53da/pr53da-dss-0024/rawdata/Bisulfite-Seq/RRBS
workdir=/dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1/work
outdir=/dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1/out

RUN=$1

# First we need to sort the bam files by reference position with picard
picard SortSam I=${workdir}/${RUN}_trimmed_bismark_bt2.bam O=${workdir}/${RUN}_trimmed_bismark_bt2_sort.bam SORT_ORDER=coordinate

# Then we will add read groups to sorted bam files
## 1) extract sample name and read group name
fbname=$(basename ${workdir}/${RUN}_trimmed_bismark_bt2_sort.bam | sed -E &#39;s/(([^_]*_){4}).*/\1/; s/_$//&#39; )
frgname=$(basename ${workdir}/${RUN}_trimmed_bismark_bt2_sort.bam | sed -E &#39;s/(([^_]*_){6}).*/\2/; s/_$//&#39; )
## 2) add read groups with picard
picard AddOrReplaceReadGroups I=${workdir}/${RUN}_trimmed_bismark_bt2_sort.bam O=${workdir}/${RUN}_trimmed_bismark_bt2_sort_RG.bam ID=`basename $fbname`  LB=`basename $frgname` PL=illumina PU=`basename $frgname` SM=`basename $fbname` CREATE_INDEX=true VALIDATION_STRINGENCY=SILENT SORT_ORDER=coordinate

# call SNPs with biscuit, compress and tabix index them
biscuit pileup -o ${workdir}/${RUN}_trimmed_bismark_bt2_sort_RG_biscuit.vcf ${genome} -o 30
${workdir}/${RUN}_trimmed_bismark_bt2_sort_RG.bam | bgzip ${outdir}/${RUN}_trimmed_bismark_bt2_sort_RG_biscuit.vcf | tabix -p vcf ${outdir}/${RUN}_trimmed_bismark_bt2_sort_RG_biscuit.vcf

</code></pre>
<p>And create a bash script for each sample:</p>
<pre><code>for sample in $(cat /dss/dsslegfs01/pr53da/pr53da-dss-0024/projects/2021_EpigeneticsTits/Bisulfate_Conversion/Genomes/Parus.major.v1.1/Libraries.list); do sbatch -J ${sample} RRBS_SNPcall.sh ${sample}; done </code></pre>
<h2 id="merging-and-filtering">Merging and Filtering</h2>
<pre><code>
#first, merge all vcf files together
bcftools merge *.vcf.gz | bgzip &gt; parus.raw.snps.vcf.gz

# Check the total number of SNPs before filtering
zcat parus.raw.snps.vcf.gz | grep -v &#39;#&#39; | print &#39;Number of SNPs before filtering is:&#39; wc -l
 

#keep only sites with a maf 0.05, genotype min depth 10 and max depth 150, shared between 50% of samples

vcftools --gzvcf raw.snps.vcf.gz --recode --recode-INFO-all --bed ${chrsbed} --maf 0.05 --minDP 10 --maxDP 150 --max-missing .5 --out dp10.maf05.mm5.snps

# Check the total number of SNPs after filtering
zcat dp10.maf05.mm5.filt.snps.vcf.gz| grep -v &#39;#&#39; | print &#39;Number of SNPs after filtering is:&#39; wc -l

</code></pre>
<h1 id="kinship-matrix">Kinship matrix</h1>
<p>Calculate kinship matrix with vcftools</p>
<pre><code>
vcftools --gzvcf dp10.maf05.mm5.filt.snps.vcf.gz --relatedness2
</code></pre>

</main>


</body>
</html>
